src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> newNode;
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            newNode = new Node<>(null, value, first);
+            first.prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new RuntimeException("Could not add list, because it is null");
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isEmpty()) {
+            throw new IndexOutOfBoundsException("List is Empty!");
+        }
+        Node<T> deleteNode = getNodeByIndex(index);
+        unlink(deleteNode);
+        return deleteNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+
+        Node<T> current = first;
+        for (int i = 0; i < size; i++) {
+            if (isEqual(current.item, object)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index, size);
+        Node<T> result = first;
+        if ((size / 2) > index) {
+            for (int i = 0; i < index; i++) {
+                result = result.next;
+            }
+        } else {
+            result = last;
+            for (int i = size - 1; i > index; i--) {
+                result = result.prev;
+            }
+        }
+        return result;
+    }
+
+    private void unlink(Node<T> deleteNode) {
+        if (size == 1) {
+            first = last = null;
+        } else if (deleteNode == first) {
+            first = deleteNode.next;
+            first.prev = null;
+        } else if (deleteNode == last) {
+            last = deleteNode.prev;
+            last.next = null;
+        } else {
+            deleteNode.prev.next = deleteNode.next;
+            deleteNode.next.prev = deleteNode.prev;
+        }
+        size--;
```

Comment:
Move `size--` in the top of method and call `isEmpty` instead of `size == 1`

Reply:
done)) thank you for your comments))